### ISSUE TITLE
Running dbt w/ multiple threads on Redshift slow due to exclusive lock

### ISSUE THREAD

**Issue by Beetelbrox** [dbt-labs/dbt#3118]
### Describe the bug
This issues is similar to #3112, but not limited to only RA3 nodes or views.

We are running our dbt pipeline in our company's quite loaded Redshift cluster with 'regular' DC2 nodes, where DDL may take quite a bit of time to run especially on Monday mornings. With the intention of reducing our lead time we bumped up the number of dbt threads to run more of our base relations (which included views, tables, incrementals) in parallel as they were the main bottleneck. By doing this we found out not only that we didn't get much of a speedup, but also we observed a quite interesting behaviour on the running times of the models, where a large amount of models started at the same time but finished in a sequential fashion (picture below). There was also no correlation between the number of rows/table size and the running time here, some of the slowest tables had less than 10 fields and a couple thousand rows, and incremental tables happened to consistently be the fastest despite usually having to load more data than regular tables (and of course views).

Digging a bit deeper into this we found out in Redshift logs that there was a huge time gap between any 'DROP' transaction and the previous one in the same model. By checking Redshift connection's implementation we found out that class [RedshiftConnectionManager](https://github.com/fishtown-analytics/dbt/blob/49b8693b1111c34fc3b134237b0f2ef7046582bf/plugins/redshift/dbt/adapters/redshift/connections.py#L67) implements an process-level exclusive lock to avoid DROP ... CASCADE conflicts, which essentially prevents two models running in different threads as part of the same `dbt run` to execute DROP transactions at the same time. Because running a table/view model requires several DROP operations to clean backup and temporary relations and as no two models can DROP at the same time because of the exclusive lock, it seems that models running in parallel can (very often actually) spend a significant amount of time competing among themselves to acquire the lock several times, which would explain the pattern in the model execution finish times. Also, because incremental models require fewer DROPs, it would explain why they finish faster.

To test this we ran each model in parallel in separate `dbt run` commands (using Airflow) and the issue disappeared completely, with models running much faster and taking longer when they had to move more data.

### Steps To Reproduce
On a Redshift cluster execute `dbt run` in multiple threads in a project with many relations. Might require a relatively large number of threads or a relatively loaded cluster to observe it properly.

### Expected behavior
Most relations finish at roughly the same time, taking longer depending on the table size

### Potential workarounds
In our case we dealt with this by dramatically reducing the number of table/views and using more ephemeral models, but this is not ideal (forced us to make our pipeline much leaner though) as we still can observe this happening especially at the beginning of the pipeline where more models are running in parallel. As the lock exists to prevent concurrent delete issues I'm not sure how this could be averted, but an override option would be nice for 'safe' sections of the pipeline that could run without the lock (for example all our views are late binding and we do not have additional objects depending on the pipeline relations)

### Screenshots and log output
Model execution time gantt chart obtained by parsing the output of `dbt run` when running multiple threads. x-axis is time.
![image](https://user-images.githubusercontent.com/9376816/108590004-97c19400-7361-11eb-9d5a-d60d7a5b8c4f.png)

### System information
**Which database are you using dbt with?**
- [ ] postgres
- [x] redshift
- [ ] bigquery
- [ ] snowflake
- [ ] other (specify: ____________)

**The output of `dbt --version`:**
```
installed version: 0.18.1
   latest version: 0.19.0

Plugins:
  - bigquery: 0.18.1
  - postgres: 0.18.1
  - redshift: 0.18.1
  - snowflake: 0.18.1
```

**The operating system you're using:**
Ubuntu 20.04

**The output of `python --version`:**
Python 3.7.5

### Additional context
N/A

**Comment by jtcohen6** [comment]
@Beetelbrox Thanks for the detailed description. Overall I agree with your assessment that this seems closely related to the one in #3112, and it's quite possible that workarounds (= hacks) we develop for one issue may serve to ameliorate the other.

> To test this we ran each model in parallel in separate `dbt run` commands (using Airflow) and the issue disappeared completely, with models running much faster and taking longer when they had to move more data.

This is the bit I find surprising. Why would dbt's lock-and-drop transactions compete only when running across multiple threads within the same invocation, but not when running across multiple different invocations (as triggered by Airflow)?

> As the lock exists to prevent concurrent delete issues I'm not sure how this could be averted, but an override option would be nice for 'safe' sections of the pipeline that could run without the lock (for example all our views are late binding and we do not have additional objects depending on the pipeline relations)

Yes, I think a possible resolution here is for dbt to differentiate between relations requiring a safe drop transaction—because they have bound dependent views, or because someone could be querying them concurrently—and relations that have no dependencies, are unlikely to be queried, and can be dropped simply (no `cascade`) and without opening a new locking transaction.

The proposed remedy for #3112 is to try this for the temp and backup relations, which are necessary for the atomic rename-swap materialization logic but not used for anything else. While not removing the need for locking drop-cascade transactions entirely, it would be a definitely improvement.

> Model execution time gantt chart obtained by parsing the output of `dbt run` when running multiple threads. x-axis is time.

This is a really cool visualization! Could you help me understand what exactly you're seeing here? Are you in fact running dbt with dozens of threads? While it depends on your cluster / WLM settings, I know that Redshift only supports a set number of concurrent connections, so many of those may be queued/delayed for other reasons.

I'm happy to keep this issue open, but at this point I'm reading it as a clarification of the impact of #3112, rather than a recasting of the problem or a reason to think differently about proposed solutions (except that both `view` and `table` materializations are implicated). Are you thinking about this differently?

**Comment by Beetelbrox** [comment]
@jtcohen6 Thanks!

>This is the bit I find surprising. Why would dbt's lock-and-drop transactions compete only when running across multiple threads within the same invocation, but not when running across multiple different invocations (as triggered by Airflow)?

As far as I can tell, the drop_lock in dbt is implemented as a [process-wide lock](https://github.com/fishtown-analytics/dbt/blob/6c6649f9129d5d108aa3b0526f634cd8f3a9d1ed/plugins/redshift/dbt/adapters/redshift/connections.py#L18) using the python multiprocessing package. Unless I'm missing some detail on how it works, this has the effect of different threads in the same invocation (aka process) will be competing for acquiring the parent process' single lock. Running each model in separate `dbt run`s (eg via airflow) concurrently will spawn different processes with a different lock for each one, hence the different behaviour.

>Yes, I think a possible resolution here is for dbt to differentiate between relations requiring a safe drop transaction—because they have bound dependent views, or because someone could be querying them concurrently—and relations that have no dependencies, are unlikely to be queried, and can be dropped simply (no cascade) and without opening a new locking transaction.

The proposed remedy for #3112 is to try this for the temp and backup relations, which are necessary for the atomic rename-swap materialization logic but not used for anything else. While not removing the need for locking drop-cascade transactions entirely, it would be a definitely improvement.

Indeed, we've observed that every drop takes a while so every bit helps!

> This is a really cool visualization! Could you help me understand what exactly you're seeing here? Are you in fact running dbt with dozens of threads? While it depends on your cluster / WLM settings, I know that Redshift only supports a set number of concurrent connections, so many of those may be queued/delayed for other reasons.

Oops, sorry for the lack of clarification here. This is a Gantt chart of the dbt model execution time for each one of our models executed concurrently, x axis is time, y axis are the models. Timestamps for start and end were parsed out from the dbt run output. We can see how we have a bunch of models that start at the same time and some of them take much longer to load than others, the interesting part here is that the longer-running models are all small table models while the fastest ones are incrementals that have to load much more data at each incremental batch.
Regarding the level of concurrency, we found out that it was taking a long time just to run all of them with 4/8 threads, so we tried with 1 thread/source model (around 50) and then observed this behaviour. Just for the records, for the rest of the pipeline we use 8 threads.
Regarding the WLM & other redshift workload configs we initially thought them to be the source of the slowness, but we checked and we had 0 time in the various queues. All that Redshift registered was a very long gap between each DROP transaction submission and the previous one.

> I'm happy to keep this issue open, but at this point I'm reading it as a clarification of the impact of #3112, rather than a recasting of the problem or a reason to think differently about proposed solutions (except that both view and table materializations are implicated). Are you thinking about this differently?

Indeed, at the moment of the initial writing the conversation in #3112 had not yet progressed to the lock situation and was only restricted to views, so I wanted add our findings too. Feel free to close it if appropriate. Thanks a lot!
